### ISSUE TITLE
Very strange never constructed warning

### ISSUE THREAD

**Issue by ogarcia** [rust-lang/rust#151447]
### Code

The full source code is here: https://gitlab.com/ogarcia/lesma/

First of all, I apologize if this isn't really a bug, but personally, I think it is, since I can't find any reason why this warning might appear.

I have this struct in `models.rs`:
```rust
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(crate = "rocket::serde")]
pub struct LesmaMeta {
    pub hash: String,
    pub password: Option<String>,
    pub name: String,
    pub binary: bool,
    pub limit: u8,
    pub current: u8,
    #[serde(with = "time::serde::rfc3339")]
    pub create: OffsetDateTime,
    #[serde(with = "time::serde::rfc3339")]
    pub expire: OffsetDateTime
}
```

And the struct is used in `lib.rs` for example in this function:
```rust
    /// Saves a plain text lesma
    pub fn save_plain(&self, limit: u8, expire: u32, password: Option<String>, lesma: &String) -> Result<(), Error> {
        // Calculate BLAKE3 hash of lesma
        let lesma_hash = blake3::hash(lesma.as_bytes()).to_hex().to_string();
        // Get data directory and counter, data and meta files
        let data_dir = self.storage.join("data").join(&lesma_hash);
        let counter_file = data_dir.join("_counter");
        let data_file = data_dir.join("_data");
        let meta_file = self.storage.join("meta").join(&self.hash);
        // Fill in the metadata
        let lesma_meta = LesmaMeta {
            hash: lesma_hash,
            password: password.and_then(|p| if p.is_empty() {
                None
            } else {
                Some(blake3::hash(p.as_bytes()).to_hex().to_string())
            }),
            name: "lesma".to_string(),
            binary: false,
            limit,
            current: 0,
            create: OffsetDateTime::now_utc(),
            expire: OffsetDateTime::now_utc() + Duration::hours(expire.into())
        };
        // Serialize metadata in JSON
        let meta_json = json::to_string(&lesma_meta)?;
        // Stores the metadata in the metadata file
        write(meta_file, meta_json)?;
        // Check if the data already exists
        if data_dir.is_dir() {
            // Add +1 to counter file
            let mut counter = read_to_string(&counter_file)?.trim().parse::<u32>().map_err(|e| Error::new(ErrorKind::InvalidData, e))?;
            counter += 1;
            write(counter_file, counter.to_string())?;
        } else {
            // Create new data directory
            create_dir(&data_dir)?;
            // Save data
            write(data_file, lesma)?;
            write(counter_file, "1")?;
        }
        Ok(())
    }
```

I expected to see this happen: nothing

Instead, this happened:
```
warning: struct `LesmaMeta` is never constructed
  --> src/models.rs:33:12
   |
33 | pub struct LesmaMeta {
   |            ^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

warning: `lesma` (bin "lesma") generated 1 warning
```

### Version it worked on

It most recently worked on: Rust 1.89

### Version with regression

Since Rust 1.90

But I'm using 1.92:
`rustc --version --verbose`:
```
rustc 1.92.0 (ded5c06cf 2025-12-08)
binary: rustc
commit-hash: ded5c06cf21d2b93bffd5d884aa6e96934ee4234
commit-date: 2025-12-08
host: x86_64-unknown-linux-gnu
release: 1.92.0
LLVM version: 21.1.3
```

In nightly occurs too.

**Comment by reddevilmidzy** [comment]
I'm trying it now
